nova/hypervisor.go

```diff
@@ -0,0 +1,251 @@
+		if service.Host == n.hostname && service.Binary == novaComputeBinaryName {
+			if service.Status == enabledString {
+				return true, nil
+			}
+			return false, nil
+		}
+	}
+	return false, fmt.Errorf("Cannot find nova-service with hostname: %s", n.hostname)
+}
+
+func (n *Hypervisor) RefreshState() (err error) {
+	status, err:= n.hypervisorStatus()
+	if err != nil {
+		return fmt.Errorf("Cannot update hypervisor state: %v", err)
+	}
+	if status != n.Enabled {
+		logger.Info.Printf("Hypervisior status updated. New status = %v", status)
+		n.Enabled = status
+	}
+	return
+}
+
+// Disable disable node and scheduling on it.
+func (n *Hypervisor) Disable() (err error) {
+	url := n.client.ServiceURL("os-services", "disable")
+	resp, err := n.client.Request("PUT", url, gophercloud.RequestOpts{
+		JSONBody: n.body,
+		OkCodes:  []int{200, 204},
+	})
+	if err != nil {
+		return fmt.Errorf("Cannot change node state. Recieved code: %s.\nError: %v", resp.StatusCode, err)
+	}
+	logger.Info.Println("Node disabled.")
+	n.Enabled = false
+
+	return
+}
+
+// Enable change node state to enable
+func (n *Hypervisor) Enable() error {
+	url := n.client.ServiceURL("os-services", "enable")
+	resp, err := n.client.Request("PUT", url, gophercloud.RequestOpts{
+		JSONBody: n.body,
+		OkCodes:  []int{200, 204},
+	})
+	if err != nil {
+		logger.Error.Println("Cannot change node state.")
+		return fmt.Errorf("Recieved code: %s.\nError: %v", resp.StatusCode, err)
+	}
+	logger.Info.Println("Node enabled.")
+	n.Enabled = true
+
+	return nil
+}
+
+func (n *Hypervisor) isMigrated(vmID string, hostID string) (bool, error) {
+	vm := new(NovaServer)
+	url := n.client.ServiceURL("servers", vmID)
+	resp, err := n.client.Request("GET", url, gophercloud.RequestOpts{
+		OkCodes:  []int{200, 204},
+	})
+	if err != nil {
+		return false, fmt.Errorf("Cannot gather server %v information: %v", vmID, err)
+	}
+
+	if err = getJson(resp.Body, &vm); err != nil {
+		return false, fmt.Errorf("Cannot decode JSON: %v", err)
+	}
+
+	if vm.Server.HostID != hostID {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// MigrateVMs live migrate all VMs out of node
+func (h *Hypervisor) MigrateVMs() (err error) {
+	var wg sync.WaitGroup
+	if err = h.updateVMList(); err != nil {
+		return fmt.Errorf("Cannot update server list: ", err)
+	}
+
+	for _, vm := range *h.vms {
+		wg.Add(1)
+
+		go func(vmID string, hostID string) {
+			migrated := true
+			defer wg.Done()
+			for a := 1; a < liveMigrationRetry + 1; a++ {
+				er := adminactions.LiveMigrate(h.client, vmID, adminactions.LiveMigrateOpts{
+					BlockMigration: true,
+				})
+				if er.Result.Err == nil {
+					logger.Info.Printf("Attempt: %d. Request to migrate VM %s accepted\n", a, vmID)
+					migrated = false
+					break
+				}
+				logger.Warning.Printf("Attempt: %d. Cannot run migratation of VM %s: %v.\n", a, vmID, er.Result.Err)
+				time.Sleep(time.Duration(a * 10) * time.Second)
+			}
+			for counter := 0; !migrated ; counter++ {
```

Comment:
We should enter this loop only if we know that live migrate request was accepted.